File: java/day04/Demo01.java
package day04;

/*
面向对象三大特征
封装：把属性编程私有的，提供set/get方法访问
继承：
多态：
函数的封装和调用
 */
public class Demo01 {
    public static void main(String[] args) {
        Student xiaoHua = new Student();
        xiaoHua.name = "小花";//实例.属性，设置属性的值
        xiaoHua.age = 18;
        xiaoHua.gender = "女";
        System.out.println(xiaoHua.name);//获取属性的值
        System.out.println(xiaoHua.toString());

        Student1 xiaoMei = new Student1();
//        xiaoMei.name = "";//私有属性，类外部无法访问
        xiaoMei.setName("小美");//设置属性的值
        System.out.println(xiaoMei.getName());
        System.out.println(xiaoMei.toString());
        xiaoMei.setGender("女");
        System.out.println(xiaoMei.toString());
        xiaoMei.setGender("鬼");//设置非法的值，
        System.out.println(xiaoMei.toString());
        xiaoMei.getGender();
        System.out.println(xiaoMei.toString());
        xiaoMei.setAge(-1);//设置非法年龄
        System.out.println(xiaoMei.toString());


    }
}

class Student {
    String name;
    int age;
    String gender;

    @Override
    public String toString() {
        return "Student{" +
                "name='" + name + '\'' +
                ", age=" + age +
                ", gender='" + gender + '\'' +
                '}';
    }
}

class Student1 {
    //使用private修饰，只能在类的内部访问。
    private String name;
    private int age;
    private String gender;

    //set方法，获取属性的值
    public void setName(String n) {
        name = n;
    }

    //get方法
    public String getName() {
        return name;
    }

    public void setAge(int a) {
        if (age > 0) {
            age = a;
        } else {
            System.out.println("请输入正确的年龄");
        }
    }

    public int getAge() {
        return age;
    }

    public String getGender() {
        return gender;
    }

    public void setGender(String g) {
        if (g.equals("男") || g.equals("女")) {
            gender = g;
        } else {
            System.out.println("参数输入错误，请输入男/女");
        }
    }

    @Override
    public String toString() {
        return "Stdent1{" +
                "name='" + name + '\'' +
                ", age=" + age +
                ", gender='" + gender + '\'' +
                '}';
    }
}

class Student2 {
    //私有
    private String name;
    private int age;
    private String gender;

    //get/set可以自动生成
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public int getAge() {
        return age;
    }

    public void setAge(int age) {
        this.age = age;
    }

    public String getGender() {
        return gender;
    }

    public void setGender(String gender) {
        this.gender = gender;
    }
}

File: python/python/day04/demo01.py
'''
面向对象编程 Object Oriented Programming - OOP
对象： 万事万物皆对象
类：某一类具有相同属性的对象的抽象
面向对象三大特性：封装，继承，多态
'''

class Student: # 类名首字母大写

    # 方法：写在类里面的函数(function)，称为方法method
    # 构造方法 constructor
    def __init__(self,name,score,gender,height=175): # 方法的第一个参数必须是self
        # self.xxx 就是属性，也是成员变量
        # name 局部变量
        # __xxx 变量前面加__,就变成了私有变量 private,只能在类的内部访问，外部无法访问
        self.__name = name
        self.__score = score
        self.__gender = gender
        self.height = height

    # 方法
    def study(self):
        print('%s正在学习...'%self.__name)

    def show(self):
        print('%s的考试成绩是：%d'%(self.__name,self.__score))

    '''
        封装
        首先将成员变量私有化(self.__xxx)
        然后通过添加get/set方法来访问私有变量
    '''
    # 获取self.__name属性的值
    def get_name(self):
        return self.__name

    # 修改self.__name属性的值
    def set_name(self,name):
        self.__name = name

    def get_score(self):
        return self.__score

    def set_score(self,score):
        if score>=0 and score<=100:
            self.__score = score
        else:
            # 抛出一个异常
            raise ValueError('非法的成绩！')

    # 性别对应的get/set 方法
    def get_gender(self):
        return self.__gender

    def set_gender(self,gender):
        # if gender=='M' or gender=='F':
        if gender in ['M','F']:
            self.__gender = gender
        else:
            raise ValueError('性别输入错误')

# 调用构造方法创建对象
tom = Student('Tom',90,'M') # 实际调用的就是该类的__init__()方法
julia = Student('Julia',88,'F')

tom.study()
julia.study()

print(tom)
print(julia)
print(Student)

# tom.__name = 'Tommy'
tom.set_name('Tommy')
tom.study()

tom.show()
# tom.__score = 95
tom.show()

julia.show()

# tom.__score = -5
# tom.set_score(500)
tom._Student__score = -100
tom.show()
